packages/core/contracts/zodiac/OptimisticOracleModule.sol

```diff
@@ -0,0 +1,197 @@
+
+    event ProposalDeleted(uint256 indexed proposalId);
+
+    FinderInterface public finder;
+    SkinnyOptimisticOracleInterface optimisticOracle;
+
+    IERC20 public collateral;
+    uint256 public bond;
+    string public rules;
+    uint256 public liveness;
+    bytes32 public identifier = "ZODIAC";
+
+    struct Transaction {
+        address to;
+        uint256 value;
+        bytes data;
+    }
+
+    struct Proposal {
+        Transaction[] transactions;
+        uint256 requestTime;
+        bytes ancillaryData;
+        bool status;
+    }
+
+    Proposal[] public proposals;
+
+    /**
+     * @notice Construct Optimistic Oracle Module.
+     * @param _finder Finder address.
+     * @param _owner Address of the owner.
+     * @param _collateral Address of the ERC20 collateral used for bonds.
+     * @param _bond Bond required (must be at least as large as final fee for collateral type).
+     * @param _rules Reference to the rules for the Gnosis Safe (e.g., IPFS hash or URI).
+     */
+    constructor(
+        address _finder,
+        address _owner,
+        address _collateral,
+        uint256 _bond,
+        string memory _rules
+    ) {
+        bytes memory initializeParams = abi.encode(_finder, _owner, _collateral, _bond, _rules);
+        setUp(initializeParams);
+    }
+
+    function setUp(bytes memory initializeParams) public override initializer {
+        __Ownable_init();
+        (address _finder, address _owner, address _collateral, uint256 _bond, string memory _rules) =
+            abi.decode(initializeParams, (address, address, address, uint256, string));
+        finder = FinderInterface(_finder);
+        optimisticOracle = _getOptimisticOracle();
+        // check collateral is whitelisted
+        collateral = IERC20(_collateral);
+        // check bond amount is large enough
+        bond = _bond;
+        rules = _rules;
+        setAvatar(_owner);
+        setTarget(_owner);
+        transferOwnership(_owner);
+
+        emit OptimisticOracleModuleDeployed(_owner, avatar, target);
+    }
+
+    function setBond(uint256 _bond) public onlyOwner {
+        // value of the bond required for proposals (must be greater than optimistic oracle final fee)
+        bond = _bond;
+    }
+
+    function setCollateral(IERC20 _collateral) public onlyOwner {
+        // ERC20 token to be used as collateral (must be approved by UMA Store contract).
+        collateral = _collateral;
+    }
+
+    function setRules(string memory _rules) public onlyOwner {
+        // set reference to the rules for the avatar (e.g. an IPFS hash or URI)
+        rules = _rules;
+    }
+
+    function setLiveness(uint256 _liveness) public onlyOwner {
+        // set liveness for disputing proposed transactions
+        liveness = _liveness;
+    }
+
+    function proposeTransactions(Transaction[] memory transactions, bytes memory ancillaryData) public {
```

Comment:
you're going to want to add some reentrancy guards to this public methods that make external calls